File: naru_agent/knowledge/hybrid_store.py
"""Hybrid knowledge store that merges results from multiple stores."""

from __future__ import annotations

from naru_agent.knowledge.base import BaseKnowledgeStore, KnowledgeResult


class HybridKnowledgeStore(BaseKnowledgeStore):
    """Merge search results from multiple knowledge stores with weighted scoring.

    Args:
        stores: List of knowledge stores to query.
        weights: Per-store weight multipliers (default: equal weights).
    """

    def __init__(
        self,
        stores: list[BaseKnowledgeStore],
        weights: list[float] | None = None,
    ) -> None:
        if not stores:
            raise ValueError("At least one store is required")
        if weights and len(weights) != len(stores):
            raise ValueError("weights must match the number of stores")
        self._stores = stores
        self._weights = weights or [1.0] * len(stores)

    def search(self, query: str, top_k: int = 3) -> list[KnowledgeResult]:
        """Search all stores and return merged, re-scored results."""
        all_results: list[KnowledgeResult] = []

        for store, weight in zip(self._stores, self._weights):
            # Request more results per store for better diversity after merging
            per_store_k = top_k * len(self._stores)
            results = store.search(query, top_k=per_store_k)
            for r in results:
                all_results.append(
                    KnowledgeResult(
                        text=r.text,
                        score=r.score * weight,
                        metadata=r.metadata,
                    )
                )

        all_results.sort(key=lambda r: r.score, reverse=True)
        return all_results[:top_k]

File: js/src/knowledge/hybrid-store.ts
import type { BaseKnowledgeStore, KnowledgeResult } from "./base.js";
import { formatKnowledgeContext } from "./base.js";

/**
 * Multi-source knowledge store that merges results from multiple stores
 * with weighted scoring.
 */
export class HybridKnowledgeStore implements BaseKnowledgeStore {
  private stores: BaseKnowledgeStore[];
  private weights: number[];

  constructor(config: {
    stores: BaseKnowledgeStore[];
    weights?: number[];
  }) {
    this.stores = config.stores;
    this.weights =
      config.weights ?? config.stores.map(() => 1 / config.stores.length);
  }

  async search(query: string, topK = 3): Promise<KnowledgeResult[]> {
    const allResults = await Promise.allSettled(
      this.stores.map((store, i) =>
        store.search(query, topK).then((results) =>
          results.map((r) => ({
            ...r,
            score: r.score * this.weights[i],
          })),
        ),
      ),
    );

    const merged: KnowledgeResult[] = [];
    const seen = new Set<string>();

    for (const result of allResults) {
      if (result.status === "fulfilled") {
        for (const r of result.value) {
          // Deduplicate by text content
          const key = r.text.slice(0, 100);
          if (!seen.has(key)) {
            seen.add(key);
            merged.push(r);
          }
        }
      }
    }

    return merged.sort((a, b) => b.score - a.score).slice(0, topK);
  }

  formatContext(results: KnowledgeResult[], minScore = 0.3): string {
    return formatKnowledgeContext(results, minScore);
  }
}
